annotation/functions.scm

```diff
@@ -121,13 +121,114 @@
    ; res is list of the GO terms directly related to 
    ; the input gene (g) that are members of the input namespaces
    (define res (find-memberln g namespaces))
-
+   (define go-regulates (append-map (lambda (go) (find-go-plus go regulates part-of)) res))
    (define all-parents (loop num-parents res '()))
 
-   (cons (node-info g) all-parents)
+   (append (node-info g) all-parents go-regulates)
+)
+
+(define regulates-rln (List (Concept "GO_regulates") (Concept "GO_positively_regulates") (Concept "GO_negatively_regulates")))
+
+(define (do-find-regulates input-atom)
+      (append (append-map (lambda (reg) (run-query (Bind 
+            (Evaluation
+               (Predicate (cog-name reg))
+               (ListLink
+                  (gar input-atom)
+                  (Variable "$go")
+               )
+            )
+
+            (Evaluation
+               (Predicate (cog-name reg))
+               (ListLink
+                  (gar input-atom)
+                  (Variable "$go")
+               )
+            )
+
+         ))) (cog-outgoing-set (gdr input-atom)))
+
+         (append-map (lambda (reg) (run-query (Bind 
+            (Evaluation
+               (Predicate (cog-name reg))
+               (ListLink
+                  (Variable "$go")
+                  (gar input-atom)
+               )
+            )
+
+            (Evaluation
+               (Predicate (cog-name reg))
+               (ListLink
+                  (Variable "$go")
+                  (gar input-atom)
+               )
+            )
+
+         ))) (cog-outgoing-set (gdr input-atom))))
```

Comment:
There is no need to search the reverse https://github.com/MOZI-AI/annotation-scheme/pull/194/files#diff-aed3d7e3a5790c8e073ca5611e02a67fR152-R169

Here is example data:
a -> regulates -> b
c -> regulated_by -> a

imported as:
Evaluation predicate "regulates" Concept "a" Concept "b"
Evaluation predicate "regulates" Concept "a" Concept "c"

When you search for go-terms a regulates, it should be straight forward,  you give "regulates" predicate and get "b" and "c"

Reply:
Well, `Evaluation predicate "regulates" Concept "a" Concept "c"` seems wrong because in your example a doesn't regulate c, it is c that regulates a. No?

Reply:
`c **regulated_by** a `means `a **regulates** c`

Reply:
Also, I think we agreed on a meeting that if a user selects to include regulates relationships,  for each GO term that the input gene is member, we will include both the GO terms that it regulates and it is regulated by

Reply:
> `c **regulated_by** a `means `a **regulates** c`

Oh, you are right!

How about the cases like this

```
 a regulates b
 c regulates a
```

We want the function to return both `b` and `c`, right?

Reply:
There should not be any **regulated_by** or **negatively/positively regulated_by** or **part of** relationships in the atomspace as I have fixed the dataset import script to make them equivalent to **regulates** **negatively/positively regulates** and **has part** respectively, in a reverse order, as i show you above in the example.

Reply:
> > `c **regulated_by** a `means `a **regulates** c`
> 
> Oh, you are right!
> 
> How about the cases like this
> 
> ```
>  a regulates b
>  c regulates a
> ```
> 
> We want the function to return both `b` and `c`, right?

if you are searching for go terms that **a** regulates, then it should only be **b**

Reply:
Not to confuse the issue but what about "upregulates" and "downregulates"?  viz `a upregulates b` means b is expressed more, whenever a is expressed more.  Whereas `a downregulates b` means that  b is expressed less, whenever a is expressed more.  This promotion vs. suppression  is presumably interesting/important (and I presume it is what "positively/negatively" was meant to be)

Reply:
> if you are searching for go terms that a regulates, then it should only be b
 having c in the result is incorrect, because it doesnt get regulated by a, rather it regulates a

Didn't we agree to remove the `regulated_by` relationships and run a bi-directional search when doing the annotation? So I don't see why returning `c` is incorrect

Reply:
what we agreed is to remove the `regulated_by` relationships and make them `regulates `relationship by reversing the order, as for the PM search, it will only look for one way.

Reply:
@mjsduncan can you weigh in on this?

Reply:
there are use cases for both possibilities.  if we're looking for causal relationships for instance, we don't want *c*_but if we are doing a general annotation or bio-atomspace subgraph generation we want *c*.  @linas i think `up_` and `down_` `regulates` are being treated the same as regulates but should also inherit from `regulates`.

Reply:
So should we keep the current code (which returns `b` & `c`) or should I modify it to not return `c`?